executor/set.go

```diff
@@ -92,7 +92,12 @@ func (e *SetExecutor) Next(ctx context.Context, req *chunk.Chunk) error {
 				if err1 != nil {
 					return err1
 				}
-				sessionVars.Users[name] = fmt.Sprintf("%v", svalue)
+
+				if len(value.Collation()) != 0 {
+					sessionVars.Users[name] = types.NewCollationStringDatum(svalue, value.Collation(), collate.DefaultLen)
+				} else {
+					sessionVars.Users[name] = types.NewStringDatum(svalue)
```

Comment:
The collation  shouldn't be "", is this branch necessary?

Reply:
for sql `set @test_collate_var = 1`, collation will be "".
In mysql, `collation(test_collate_var )` return `binary`.
May be I should handle `TODO: we should consider the type of the argument, but not take it as string for all situations.` In `func (c *getVarFunctionClass) getFunction`

Reply:
We don't need to consider the type of the variable. Take it as string is enough.
What we need to do here is convert the string into a datum, the charset and collation can be got by `ctx.GetSessionVars().GetCharsetInfo()`
And your PR is almost complete,  please don't give up this pr.

Reply:
ok...Sorry for my  foolish operation.
I will open a new pr for it

Reply:
The implement of collation has many implications，feel free to ask me if you are not sure.